July-Morning/assignment2/Assignment_2.cpp

```diff
@@ -0,0 +1,166 @@
+#include <iostream>
+#include <string>
+#include <list>
+
+template <typename T> struct node
+{
+    T key;
+    node * left;
+    node * right;
+    node (T init_key)
+    {
+        key = init_key;
+        left = NULL;
+        right = NULL;
+    }
+};
+
+// Here we assume that all keys in the tree are unique
+template <typename T> class binary_tree 
+{
+    public:
+    
+        node<T> * root;
+    
+        binary_tree() : root(NULL) {}
+        
+        ~binary_tree()
+        {
+            if (root != NULL)
+                destroy(root);
+        }
+        
+        void destroy (node<T> * s_root) 
+        {
+            if (s_root->left != NULL)
+                destroy (s_root->left);
+            if (s_root->right != NULL)
+                destroy (s_root->right);
+            delete s_root;   
+        }
+        
+        bool is_child (T& key, node<T> * parent) 
```

Comment:
I didn't fix that because from my point of view such change will make final checks in find_lowest_common_ancestor function more complex and less readable. Also I am not sure that "a node being a child of itself" definition is intuitive enough, though it will definitely save me from code duplication and make it shorter. Am I wrong? What option is better here?

Reply:
This choice is OK, this is a helper function and its logic should support the main code. If you think that simplifying this will complicate the code in total then there is no need to, this is a good reason.

About the semantics:
A "child" is a direct descendant, (so a node is really not a child to self...), but you changed the name of this function to "is_descendant", and a node can be its own descendent which is a more general definition... If you think that is not clear enough, then "is_in_subtree" should remove all doubt (a node and all its descendants are in its subtree).

About the algorithm:
In the assignment, the common ancestor of a node and its descendant should be the top node.
For example. in the following tree:
```
      7
     / \
    3   4
   / \   \
  2   5   8
 / \
1   6
```
The common ancestor of 7 and 2 is 7; the common ancestor of 5 and 3 is 3.
Is that the case in your solution?
From reading your code (I did not run it so I am not sure), I am not sure this is the case.
I think that fixing the `is_descendent` function to count a node as a descendent of itself will solve this without adding checks in `find_lowest_common_ancestor`.

Reply:
I see, it is a bit contradictory issue (can a node be considered its own descendant/ancestor or not), we discussed it today. For now I will just put a comment to this function.
Thank you for such a detailed exaplanation!